File: src/main/java/tech/simter/jackson/javatime/support/JavaTimeConfiguration.java
package tech.simter.jackson.javatime.support;

import com.fasterxml.jackson.databind.Module;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.autoconfigure.jackson.JacksonAutoConfiguration;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import tech.simter.jackson.javatime.JavaTimeFormat;
import tech.simter.jackson.javatime.JavaTimeModule;

import java.time.temporal.TemporalAccessor;
import java.util.HashMap;
import java.util.Map;

/**
 * An auto configuration from spring-boot.
 *
 * @author RJ
 */
@Configuration
public class JavaTimeConfiguration {
  private final static Logger logger = LoggerFactory.getLogger(JavaTimeConfiguration.class);

  /**
   * Simter Global {@link JavaTimeFormat}s.
   *
   * @return custom java-time formats
   */
  @Bean("simter.jackson.java-time.global-formats")
  @ConfigurationProperties(prefix = "simter.jackson.java-time.global-formats")
  public Map<Class<? extends TemporalAccessor>, JavaTimeFormat> simterGlobalJavaTimeFormats() {
    return new HashMap<>();
  }

  /**
   * Register by method
   * {@link JacksonAutoConfiguration}.Jackson2ObjectMapperBuilderCustomizerConfiguration.StandardJackson2ObjectMapperBuilderCustomizer.configureModules].
   *
   * @param javaTimeFormats the custom java-time formats
   * @return a custom java-time module
   */
  @Bean
  public Module simterJavaTimeModule(
    @Qualifier("simter.jackson.java-time.global-formats")
      Map<Class<? extends TemporalAccessor>, JavaTimeFormat> javaTimeFormats
  ) {
    logger.debug("simter.jackson.java-time.global-formats={}", javaTimeFormats);
    JavaTimeFormat.JAVA_TIME_FORMATS.putAll(javaTimeFormats);
    logger.debug("JavaTimeFormat.JAVA_TIME_FORMATS={}", JavaTimeFormat.JAVA_TIME_FORMATS);
    return JavaTimeModule.INSTANCE;
  }
}

File: src/test/kotlin/tech/simter/jackson/ext/javatime/JavaTimeConfiguration.kt
package cn.gftaxi.webflux.dynamicdto

import com.fasterxml.jackson.annotation.JsonInclude.Include
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.Module
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import org.springframework.boot.autoconfigure.jackson.Jackson2ObjectMapperBuilderCustomizer
import org.springframework.boot.autoconfigure.jackson.JacksonAutoConfiguration
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder
import tech.simter.jackson.ext.javatime.JavaTimeModule

/**
 * @author RJ
 *
 * See spring-boot document [76.3 Customize the Jackson ObjectMapper](https://docs.spring.io/spring-boot/docs/2.0.4.RELEASE/reference/htmlsingle/#howto-customize-the-jackson-objectmapper).
 */
@Configuration
class JavaTimeConfiguration {
  /**
   * [JacksonAutoConfiguration.JacksonObjectMapperConfiguration.jacksonObjectMapper]
   */
  @Primary
  @Bean
  fun jacksonObjectMapper(builder: Jackson2ObjectMapperBuilder): ObjectMapper {
//    val objectMapper = builder.createXmlMapper(false).build<ObjectMapper>()
//    objectMapper.registerModule(JavaTimeModule())
//    return objectMapper

    val mapper = ObjectMapper()
    mapper.registerModule(JavaTimeModule())
    mapper.setSerializationInclusion(Include.NON_EMPTY) // not serialize null and empty value
    mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
    mapper.disable(DeserializationFeature.ADJUST_DATES_TO_CONTEXT_TIME_ZONE)
    mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
    mapper.configure(DeserializationFeature.ACCEPT_EMPTY_STRING_AS_NULL_OBJECT, true)
    return mapper
  }

  /**
   * Register by method
   * [JacksonAutoConfiguration.Jackson2ObjectMapperBuilderCustomizerConfiguration.StandardJackson2ObjectMapperBuilderCustomizer.configureModules]
   */
  @Bean
  fun customJavaTimeModule(): Module {
    return JavaTimeModule()
  }

  /**
   * Register by method [JacksonAutoConfiguration.JacksonObjectMapperBuilderConfiguration.jacksonObjectMapperBuilder]
   */
  @Bean
  fun customJackson(): Jackson2ObjectMapperBuilderCustomizer {
    // WebFluxConfigurer
    return Jackson2ObjectMapperBuilderCustomizer {
      // not serialize null and empty value
      it.serializationInclusion(Include.NON_EMPTY)

      it.featuresToDisable(
        DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES,
        DeserializationFeature.ADJUST_DATES_TO_CONTEXT_TIME_ZONE,
        SerializationFeature.WRITE_DATES_AS_TIMESTAMPS
      )
      it.featuresToEnable(
        DeserializationFeature.ACCEPT_EMPTY_STRING_AS_NULL_OBJECT
      )
    }
  }

  // org.springframework.boot.jackson.JsonComponentModule
  // jackson-module-parameter-names
}